ofono/src/cell-info-control.c

```diff
@@ -0,0 +1,282 @@
+				self->path, self);
+			DBG("%s created", path);
+			return &self->pub;
+		}
+	}
+	return NULL;
+}
+
+CellInfoControl *cell_info_control_ref(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self) {
+		self->refcount++;
+	}
+	return ctl;
+}
+
+void cell_info_control_unref(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && !--self->refcount) {
+		cell_info_control_free(self);
+	}
+}
+
+void cell_info_control_set_cell_info(CellInfoControl *ctl,
+				struct ofono_cell_info *ci)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && ctl->info != ci) {
+		ofono_cell_info_unref(ctl->info);
+		ctl->info = ofono_cell_info_ref(ci);
+		cell_info_control_update_all(self);
+	}
+}
+
+void cell_info_control_drop_all_requests(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self) {
+		cell_info_control_drop_all_requests_internal(self);
+		cell_info_control_update_all(self);
+	}
+}
+
+void cell_info_control_drop_requests(CellInfoControl *ctl, void *tag)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && tag) {
+		if (self->enabled &&
+			g_hash_table_remove(self->enabled, tag) &&
+			!g_hash_table_size(self->enabled)) {
+			g_hash_table_unref(self->enabled);
+			self->enabled = NULL;
+			ofono_cell_info_set_enabled(ctl->info, FALSE);
+		}
+		if (self->set_interval &&
+			g_hash_table_remove(self->set_interval, tag)) {
+			int ms;
+
+			if (g_hash_table_size(self->set_interval)) {
+				ms = cell_info_control_get_interval(self);
+			} else {
+				g_hash_table_unref(self->set_interval);
+				self->set_interval = NULL;
+				ms = INT_MAX;
+			}
+			ofono_cell_info_set_update_interval(ctl->info, ms);
+		}
+	}
+}
+
+void cell_info_control_set_enabled(CellInfoControl *ctl,
+	void *tag, ofono_bool_t enabled)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && tag) {
+		gboolean was_enabled = (self->enabled != NULL);
+		gboolean is_enabled;
+
+		if (enabled) {
+			if (!self->enabled) {
+				self->enabled = g_hash_table_new(g_direct_hash,
+					g_direct_equal);
+			}
+			g_hash_table_insert(self->enabled, tag, tag);
+		} else if (self->enabled) {
+			g_hash_table_remove(self->enabled, tag);
+			if (!g_hash_table_size(self->enabled)) {
+				g_hash_table_unref(self->enabled);
+				self->enabled = NULL;
+			}
+		}
+
+		is_enabled = (self->enabled != NULL);
+		if (is_enabled != was_enabled) {
+			ofono_cell_info_set_enabled(ctl->info, is_enabled);
+		}
+	}
+}
+
+void cell_info_control_set_update_interval(CellInfoControl *ctl,
+	void *tag, int ms)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && tag) {
+		int old_interval = cell_info_control_get_interval(self);
+		int new_interval;
+
+		if (ms >= 0 && ms < INT_MAX) {
+			if (!self->set_interval) {
```

Comment:
Once again ofono coding style... suggests use of `!= NULL` (which personally I don't agree :). Feel free to divert from it again.